File: src/main/java/pl/allegro/experiments/chi/chiserver/domain/experiments/administration/StartExperimentCommand.java
package pl.allegro.experiments.chi.chiserver.domain.experiments.administration;

import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentDefinition;
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentsRepository;

import java.util.Objects;

public class StartExperimentCommand implements ExperimentCommand {
    private final ExperimentsRepository experimentsRepository;
    private final StartExperimentProperties startExperimentProperties;
    private final PermissionsAwareExperimentRepository permissionsAwareExperimentRepository;
    private final String experimentId;

    StartExperimentCommand(
            ExperimentsRepository experimentsRepository,
            StartExperimentProperties startExperimentProperties,
            PermissionsAwareExperimentRepository permissionsAwareExperimentRepository,
            String experimentId) {
        Objects.requireNonNull(experimentsRepository);
        Objects.requireNonNull(startExperimentProperties);
        Objects.requireNonNull(permissionsAwareExperimentRepository);
        Objects.requireNonNull(experimentId);
        this.experimentsRepository = experimentsRepository;
        this.startExperimentProperties = startExperimentProperties;
        this.permissionsAwareExperimentRepository = permissionsAwareExperimentRepository;
        this.experimentId = experimentId;
    }

    public void execute() {
        ExperimentDefinition experiment = permissionsAwareExperimentRepository.getExperimentOrException(experimentId);
        validate(experiment);
        ExperimentDefinition started = experiment.start(startExperimentProperties.getExperimentDurationDays());
        experimentsRepository.save(started);
    }

    private void validate(ExperimentDefinition experiment) {
        if (!experiment.isDraft()) {
            throw new ExperimentCommandException("Experiment is not DRAFT: " + experimentId);
        }
    }

    @Override
    public String getNotificationMessage() {
        return "was started";
    }

    @Override
    public String getExperimentId() {
        return experimentId;
    }
}

File: src/main/kotlin/pl/allegro/experiments/chi/chiserver/domain/experiments/administration/StartExperimentCommand.kt
package pl.allegro.experiments.chi.chiserver.domain.experiments.administration

import pl.allegro.experiments.chi.chiserver.application.experiments.administration.AuthorizationException
import pl.allegro.experiments.chi.chiserver.domain.UserProvider
import pl.allegro.experiments.chi.chiserver.domain.experiments.Experiment
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentId
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentsRepository


class StartExperimentCommand(private val experimentsRepository: ExperimentsRepository,
                             private val userProvider: UserProvider,
                             private val startExperimentProperties: StartExperimentProperties,
                             private val experimentId: ExperimentId) {

    fun execute() {
        val experiment = experimentsRepository.getExperiment(experimentId)
        validate(experiment)
        experimentsRepository.save(experiment!!.start(startExperimentProperties.experimentDurationDays))
    }

    private fun validate(experiment: Experiment?) {
        if (experiment == null) {
            throw ExperimentNotFoundException("Experiment not found: ${experimentId}")
        }

        val user = userProvider.getCurrentUser()
        if (!user.isOwner(experiment) && !user.isRoot) {
            throw AuthorizationException("User has no permission to edit experiment: ${experimentId}")
        }

        if (!experiment.isDraft()) {
            throw StartExperimentException("Experiment is not DRAFT: ${experimentId}")
        }

        if (startExperimentProperties.experimentDurationDays <= 0) {
            throw StartExperimentException("Experiment duration days must be greater than 0: ${experimentId}")
        }
    }
}